### PR TITLE
SAA-1009 changes to support cancellation of attendances on reciept of a prisoner being received back into prison.

### PR DESCRIPTION
**Pull request by matt-crowson-moj**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/Attendance.kt [hunk 1]
```diff
@@ -96,9 +97,9 @@ data class Attendance(
     }
 
   fun uncancel() = mark(
-    principalName = if (this.attendanceReason?.code != AttendanceReasonEnum.SUSPENDED) null else this.recordedBy,
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/Attendance.kt:99]: You should be able to cancel suspended attendances, hence the remove here.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/ActivityServiceTest.kt [hunk 2]
```diff
@@ -717,20 +771,24 @@ class ActivityServiceTest {
 
   @Test
   fun `updateActivity - prolonging the end date into the future is successful`() {
-    val activity = activityEntity(startDate = TimeSource.tomorrow(), endDate = TimeSource.tomorrow().plusDays(1))
+    val activity = activityEntity(startDate = TimeSource.tomorrow(), endDate = TimeSource.tomorrow().plusDays(1)).also { activity ->
+      activity.schedules().forEach { schedule -> schedule.slots().onEach { slot -> slot.runEveryDayOfWeek() } }
```
Comment (matt-crowson-moj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/ActivityServiceTest.kt:775]: This test was only passing if it was a Tuesday.  It should pass for any day of the week now.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/ActivityServiceTest.kt [hunk 3]
```diff
@@ -715,22 +769,27 @@ class ActivityServiceTest {
     }
   }
 
-  // @Test
+  @Test
   fun `updateActivity - prolonging the end date into the future is successful`() {
```
Comment (matt-crowson-moj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/ActivityServiceTest.kt:773]: Re-activated test.

This test was only passing if it was a Tuesday.  It should pass for any day of the week now.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/events/handlers/OffenderReceivedEventHandler.kt [hunk 4]
```diff
@@ -78,12 +80,20 @@ class OffenderReceivedEventHandler(
...
       )
-        .filter { attendance -> attendance.editable() && attendance.attendanceReason?.code == AttendanceReasonEnum.SUSPENDED }
+        .filter { attendance -> attendance.editable() && attendance.hasReason(AttendanceReasonEnum.SUSPENDED) }
         .filter { attendance ->
           (attendance.scheduledInstance.sessionDate == now.toLocalDate() && attendance.scheduledInstance.startTime > now.toLocalTime()) ||
             (attendance.scheduledInstance.sessionDate > now.toLocalDate())
         }
-        .onEach(Attendance::unsuspend)
+        .onEach { attendance ->
+          log.info("attendance ${attendance.attendanceId}")
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/events/handlers/OffenderReceivedEventHandler.kt:89]: Just noticed I left this log statement in.  Will remove in a separate PR.

**Review by iterative-it — APPROVED** [review]
LGTM

**Pull request merged**